zerver/migrations/0511_stream_creator.py

```diff
@@ -14,17 +15,25 @@ def backfill_creator_id_from_realm_audit_log(
     RealmAuditLog.STREAM_CREATED = 601
     Stream = apps.get_model("zerver", "Stream")
 
-    stream_updates = []
-    for audit_log_entry in RealmAuditLog.objects.select_related("modified_stream").filter(
-        event_type=RealmAuditLog.STREAM_CREATED,
-        acting_user_id__isnull=False,
-    ):
-        assert audit_log_entry.modified_stream is not None
-        stream = audit_log_entry.modified_stream
-        stream.creator_id = audit_log_entry.acting_user_id
-        stream_updates.append(stream)
-
-    Stream.objects.bulk_update(stream_updates, ["creator_id"], batch_size=1000)
+    batch_size = 10000
+    try:
+        for id_start in range(
+            RealmAuditLog.objects.earliest("id").id,
+            RealmAuditLog.objects.latest("id").id + 1,
```

Comment:
+1 is unsafe if multiple objects get created while the migration is running. I've been doing `+ batch_size / 2` in other migrations for this reason, since it's very unlikely that thousands of these objects are created during the migration.

Reply:
If old-style objects are still being created while the migration is running, then old-style objects are still being created after the migration finishes running, so your proposal adds no safety.

The way to safely guard against that, which I thought we had been doing, is to split the migration in two so we can (1) create the column, (2) adjust the running code to create new-style objects, and (3) backfill old-style objects while only new-style objects are being created. But that wasn’t done here.

Reply:
We generally do that type of operational work manually for Zulip Cloud deploys. Self-hosted systems are fully shut down while migrations run. 

This migration is unusual, since the backfill source in incomplete. Most similar backfill migrations are populating a nullable field, where the next migration will make it non-nullable; in that case, the next migration will fail if additional objects were created in the meantime, and once that next migration is completed, any older code will fail to create objects. For those migrations, doing a `+BATCH_SIZE/2` rather than `+1` has basically only one possible impact, which is to reduce the probability that an online backfill is incomplete, which can result in the next migration failing, requiring a few minutes of extra operational effort to rerun.

Anyway, probably what we should be doing is rewriting `docs/subsystems/schema-migrations.md` to have better examples for how to do backfills.

For this particular migration, were pre-patch code running before it was applied, those objects would end up like the older objects without a backfill data source, which is OK as a tiny race failure mode.